### PR TITLE
Introduce HttpProvider

### PR DESCRIPTION
**Pull request by jupe**
Changes:
* extract resources provider from Lockable
  * Provider (abstract)
  * ProviderList  (when resources is given as plain list of objects)
  * ProviderFile (when resources are given as file)
  * ProviderHttp (when resources are given as http url)
* support http resources provider
  * possible to give resources as http url, like http://localhost/resources

Internal changes:
* Lockable._resource_list reload list from file/http and return list of resource objects
* Constructor (re)load resources

TODO:

- [ ] Update readme

### CODE REVIEW

File: lockable/provider.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            return False
+
+    @abstractmethod
+    def _reload(self) -> None:
+        pass  # pragma: no cover
+
+    def set_resources_list(self, resources_list: list):
+        """ Load resources list """
+        assert isinstance(resources_list, list), 'resources_list is not an list'
+        self._validate_json(resources_list)
```
Comment (juhhov) [inline lockable/provider.py:64]: This should fail as _validate_json is static.

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+certifi==2021.5.30
```
Comment (juhhov) [inline requirements.txt:1]: Why separate requirements.txt when setup.py also defines dependencies?

File: lockable/provider.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+from pydash import filter_, count_by
+
+MODULE_LOGGER = logging.getLogger('lockable')
+
+
+class ProviderError(Exception):
+    """ Provider error """
+
+
+class Provider(ABC):
```
Comment (juhhov) [inline lockable/provider.py:19]: Should separate classes to own files.

File: requirements.txt (hunk 2 above)
Comment (jupe) [reply requirements.txt:1]: based on python best practises: https://packaging.python.org/discussions/install-requires-vs-requirements/ there is good to have "minimum versions" defined in `install_requires` field in `setup.py` and pinned versions in `requirements.txt`..

> Whereas install_requires requirements are minimal, requirements files often contain an exhaustive listing of pinned versions for the purpose of achieving repeatable installations of a complete environment.

File: lockable/provider.py (hunk 3 above)
Comment (jupe) [reply lockable/provider.py:19]: true, will split

**Comment by juhhov** [comment]
Ready for review?

**Comment by jupe** [comment]
there is coverage decrease but Im not sure where it comes since it doesn’t happen in my machine… have to figure out still

File: lockable/provider_http.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+            response = requests.get(uri)
+            response.raise_for_status()
+
+            # could utilise ETag or Last-Modified headers to optimize performance
+            # etag = response.headers.get("ETag")
+            # last_modified = response.headers.get("Last-Modified")
+
+            # access JSON content
+            return response.json()
+        except HTTPError as http_err:
```
Comment (juhhov) [inline lockable/provider_http.py:36]: This probably results to whole test job failing? Should this implement internal retry logic?

Comment (jupe) [reply lockable/provider_http.py:36]: I was thinking the same. Perhaps it would be good idea to retry at least some cases.

Comment (jupe) [reply lockable/provider_http.py:36]: retry logic perhaps could be introduce follow-up PR since its ”feature” and not mandatory here.

Comment (jupe) [reply lockable/provider_http.py:36]: coming here: https://github.com/jupe/py-lockable/pull/16

**Comment by juhhov** [comment]
Smoke tests of HTTP provider are positive. Noticed one possible issue though. Is it expected that `lockable --resources http://localhost:3000/resources --requirements '{"online": true}' --hostname nuc01.maas ls` leads two 4 separate get requests to server?

File: lockable/lockable.py [hunk 5]
```diff
@@ -63,68 +64,18 @@ def __init__(self, hostname=socket.gethostname(),
...
-
-    @staticmethod
-    def _validate_json(data):
-        """ Internal method to validate resources.json content """
-        counts = count_by(data, lambda obj: obj.get('id'))
-        no_ids = filter_(counts.keys(), lambda key: key is None)
-        if no_ids:
-            raise ValueError('Invalid json, id property is missing')
+    @property
+    def _resource_list(self):
```
Comment (juhhov) [inline lockable/lockable.py:76]: Calling this triggers http call each time when http provider is used.

File: lockable/lockable.py [hunk 6]
```diff
@@ -228,7 +179,6 @@ def lock(self, requirements: (str or dict), timeout_s: int = 1000) -> Allocation
         :return: Allocation context
         """
         assert isinstance(self._resource_list, list), 'resources list is not loaded'
```
Comment (juhhov) [inline lockable/lockable.py:181]: Triggers http call.

Comment (juhhov) [reply lockable/lockable.py:181]: Also `self.logger.debug("Resource list: %s", json.dumps(self._resource_list))` triggers.

**Review by juhhov — CHANGES_REQUESTED** [review]
4 http calls per allocation seems too much.

File: lockable/lockable.py (hunk 5 above)
Comment (jupe) [reply lockable/lockable.py:76]: fixed in 6f519e9

File: lockable/lockable.py (hunk 6 above)
Comment (jupe) [reply lockable/lockable.py:181]: fixed in 6f519e9

**Review by juhhov — APPROVED** [review]
Please update README.

**Pull request merged**